gcc/rust/hir/rust-hir-dump.cc

```diff
@@ -17,678 +17,2457 @@
+//   }
+//
+// If a field is an empty collection:
+//  field: empty
+//
+// If a field is optional and is currently not holding anything:
+//   field: none
 
 void
-Dump::go (HIR::Crate &crate)
+Dump::go (HIR::Crate &e)
+{
+  begin ("Crate");
+  do_inner_attrs (e);
+  do_mappings (e.get_mappings ());
+
+  visit_collection ("items", e.get_items ());
+  end ("Crate");
+}
+
+std::string
+convert_param_kind_to_str (MaybeNamedParam::ParamKind pk)
+{
+  switch (pk)
+    {
+    case MaybeNamedParam::ParamKind::UNNAMED:
+      return "UNNAMED";
+    case MaybeNamedParam::ParamKind::IDENTIFIER:
+      return "IDENTIFIER";
+    case MaybeNamedParam::ParamKind::WILDCARD:
+      return "WILDCARD";
+    }
+  gcc_unreachable ();
+}
+
+std::string
+convert_new_bind_type_to_str (UseTreeRebind::NewBindType nbt)
+{
+  switch (nbt)
+    {
+    case UseTreeRebind::NewBindType::NONE:
+      return "NONE";
+    case UseTreeRebind::NewBindType::IDENTIFIER:
+      return "IDENTIFIER";
+    case UseTreeRebind::NewBindType::WILDCARD:
+      return "WILDCARD";
+    }
+  gcc_unreachable ();
+}
+
+std::string
+convert_mut_to_str (Mutability mut)
+{
+  std::string str;
+  // FIXME this should be done elsewhere/differently, it's not specific to HIR
+  // dump.
+  switch (mut)
+    {
+    case Mutability::Imm:
+      return "Imm";
+    case Mutability::Mut:
+      return "Mut";
+    }
+  gcc_unreachable ();
+}
+
+Dump::Dump (std::ostream &stream) : stream (stream) {}
+
+/**
+ * Writes TEXT with a final newline if ENDLINE is true.
+ * If TEXT is starting the current line, its first line is indented.
+ * If TEXT is multiline, all followning lines are also indented.
+ *
+ * @param text Text to emit
+ * @param endline If true, newline is emitted after text
+ */
+void
+Dump::put (std::string text, bool endline)
 {
-  stream << "Crate {" << std::endl;
-  // inner attributes
-  if (!crate.get_inner_attrs ().empty ())
+  if (beg_of_line)
     {
-      indentation.increment ();
       stream << indentation;
-      stream << "inner_attrs: [";
-      for (auto &attr : crate.get_inner_attrs ())
-	stream << attr.as_string ();
-      stream << "]," << std::endl;
-      indentation.decrement ();
+      beg_of_line = false;
     }
 
-  indentation.increment ();
-  stream << indentation;
-  //
+  // keep multiline string indented
+  std::string::size_type pos = 0;
+  std::string::size_type prev = 0;
+  auto first = true;
+  while ((pos = text.find ('\n', prev)) != std::string::npos)
+    {
+      if (!first)
+	stream << std::endl << indentation;
+      first = false;
+
+      stream << text.substr (prev, pos - prev);
+      prev = pos + 1;
+    }
 
-  stream << "items: [";
+  if (first)
+    stream << text;
 
-  stream << indentation;
-  for (const auto &item : crate.get_items ())
+  if (endline)
     {
       stream << std::endl;
-      item->accept_vis (*this);
+      beg_of_line = endline;
     }
-  stream << std::endl;
-  stream << indentation;
+}
+
+static std::string delims[2][2] = {
```

Comment:
i think you should keep this along side the Delim enum as they are associated

Reply:
ack